### ISSUE TITLE
nnU-Netv2 pretraining and finetuning

### ISSUE THREAD

**Issue by valosekj** [ivadomed/model_seg_sci#43]
This thread summarizes experiments related to nnU-Netv2 [Pretraining and finetuning](https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/pretraining_and_finetuning.md#pretraining-on-the-source-dataset).

Note: Experiment I was moved to a separate [issue](https://github.com/ivadomed/model_seg_sci/issues/49) because it is not related to pretraining and finetuning.

**Comment by valosekj** [comment]
# Pretraining and finetuning

Since training only on subjects without metal artifacts has not improved model performance much (see the comment above), I am trying [Pretraining and finetuning](https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/pretraining_and_finetuning.md#pretraining-on-the-source-dataset).

[Terminology](https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/pretraining_and_finetuning.md#terminology):

`SOURCE_DATSET` - the dataset you intend to run the pretraining on -- only subjects without metal artifacts
`TARGET_DATASET` - the dataset you are interested in; the one you wish to fine tune on -- all subjects (including subjects with metal artifacts)

## Steps

### 1. Prepare and download data

```
mkdir sci-zurich_pretraining_and_finetuning
cd sci-zurich_pretraining_and_finetuning
git clone git@data.neuro.polymtl.ca:datasets/sci-zurich
mv sci-zurich sci-zurich_rpi
# allow ** globbing
setopt globstarshort
git annex get **/*sag_T2w*
```

### 2. Reorient to RPI

```
for image in **/*sag_T2w*.nii.gz;do echo sct_image -i $image -setorient RPI;done
```

### 3. Create `SOURCE_DATSET` -- only subjects without metal artifacts

```
python convert_bids_to_nnUNetv2.py --path-data ~/extrassd1/janvalosek/data/sci-zurich_pretraining_and_finetuning/sci-zurich_rpi --path-out ~/extrassd1/janvalosek/data/sci-zurich_pretraining_and_finetuning/sci_zurich_rpi_good_subjects_only_nnunet --dataset-name tSCILesionsZurichGoodSubjects --dataset-number 525 --split 0.8 0.2 --seed 99 --good-subjects-only --path-yaml ~/duke/projects/ml_sci_seg/QCs/qc_zurich_sagittal/qc_artifact.yml  ~/duke/projects/ml_sci_seg/QCs/qc_zurich_sagittal/qc_fail.yml
```

### 4. Create `TARGET_DATASET` -- all subjects (including subjects with metal artifacts)

```
python convert_bids_to_nnUNetv2.py --path-data ~/extrassd1/janvalosek/data/sci-zurich_pretraining_and_finetuning/sci-zurich_rpi --path-out ~/extrassd1/janvalosek/data/sci-zurich_pretraining_and_finetuning/sci_zurich_rpi_all_subjects_only_nnunet --dataset-name tSCILesionsZurichAllSubjects --dataset-number 526 --split 0.8 0.2 --seed 99
```

### 5. Create symbolic links to `$nnUNet_raw` folder

```
ln -s /home/GRAMES.POLYMTL.CA/p118175/extrassd1/janvalosek/data/sci-zurich_pretraining_and_finetuning/sci_zurich_rpi_all_subjects_nnunet/Dataset526_tSCILesionsZurichAllSubjects Dataset526_tSCILesionsZurichAllSubjects

ln -s /home/GRAMES.POLYMTL.CA/p118175/extrassd1/janvalosek/data/sci-zurich_pretraining_and_finetuning/sci_zurich_rpi_good_subjects_only_nnunet/Dataset525_tSCILesionsZurichGoodSubjects Dataset525_tSCILesionsZurichGoodSubjects
```

### 6. [Pretraining on the source dataset](https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/pretraining_and_finetuning.md#pretraining-on-the-source-dataset)

```
# TARGET_DATASET -- all subjects
nnUNetv2_plan_and_preprocess -d 526

# SOURCE_DATASET -- good subjects only
nnUNetv2_plan_and_preprocess -d 525		# this command is also necessary, although is not listed in the tutorial (https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/pretraining_and_finetuning.md#pretraining-on-the-source-dataset), othwerwise, `nnUNetv2_move_plans_between_datasets` commands fails because `nnUNetPlans.json` does not exist for 525
nnUNetv2_extract_fingerprint -d 525

# Take the plans from the target dataset and transfer it to the source
# Tutorial seems to be wrong, I had to swap 525 and 526; see [this comment](https://github.com/MIC-DKFZ/nnUNet/issues/1477#issue-1734332402)
nnUNetv2_move_plans_between_datasets -s 526 -t 525 -sp nnUNetPlans -tp nnUNetMovedPlans

# Preprocess and train
nnUNetv2_preprocess -d 525 -plans_name nnUNetMovedPlans
# `all` means https://github.com/MIC-DKFZ/nnUNet/issues/29#issuecomment-508985355
CUDA_VISIBLE_DEVICES=3 nnUNetv2_train 525 3d_fullres all -p nnUNetMovedPlans
```

**Comment by naga-karthik** [comment]
```
nnUNetv2_move_plans_between_datasets -s 526 -t 527 -sp nnUNetPlans -tp nnUNetMovedPlansSC
```

**For clarity**: This command `nnUNetv2_move_plans_between_datasets` expects the source (i.e. `-s`) to be the dataset which we are want to fine-tune on (**unlike** what is defined as "source" in the documentation). This also makes sense because we want to move the plans from the fine-tuning dataset (i.e. the one we care about) to our pre-training dataset. Likewise, the target (i.e. `-t`) for the above command is the dataset which we want to use for pre-training (again, **unlike** what's referred to as "target" in the documentation). 

A good way to remember this (and not get confused) is to know that we ultimately want the `MovedPlans` for the dataset we want to _pre-train_ on. And, this `MovedPlans` will be available only when we move the plans from the fine-tuning (`-s`) dataset to the pre-training dataset (`-t`).

**Comment by valosekj** [comment]
# Experiment IV - Pretraining and finetuning - SC and lesions - all datasets

This experiment is similar to Experiment III (https://github.com/ivadomed/model_seg_sci/issues/43#issuecomment-1628782925). But here, we use all the datasets (`sci-zurich`, `sci-colorado`, `sci-parts`).
It does pretraining on all subjects using SC masks (in order to learn context about SC anatomy), then finetuning using lesion masks. 

`SOURCE_DATSET` - the dataset you intend to run the pretraining on --  all subjects -- spinal cord masks
`TARGET_DATASET` - the dataset you are interested in; the one you wish to fine tune on -- all subjects -- lesion masks

## Steps

### 1. Download data

```
# Create a folder
mkdir all_datasets_RPI_pretraining_and_finetuning
cd all_datasets_RPI_pretraining_and_finetuning

# Clone all datasets
git clone git@data.neuro.polymtl.ca:datasets/sci-zurich
git clone git@data.neuro.polymtl.ca:datasets/sci-colorado
git clone git@data.neuro.polymtl.ca:datasets/sci-paris

# Rename datasets to make explicit that they are going to be RPI
mv sci-zurich sci-zurich_rpi
mv sci-colorado sci-colorado_rpi
mv sci-paris sci-paris_rpi

# Download relevant contrasts only (to save space)
# First, allow ** globbing
setopt globstarshort
# sci-zurich - sag_T2w
cd sci-zurich_rpi
git annex get **/*sag_T2w*
# sci-colorado - T2w (axial)
cd ../sci-colorado_rpi
git annex get **/*T2w*
# sci-paris - T2w (0.9mm iso)
cd ../sci-paris_rpi/
git annex get **/*T2w*
```

### 2. Reorient to RPI

```
cd sci-zurich_rpi
for image in **/*sag_T2w*.nii.gz;do sct_image -i $image -setorient RPI;done
cd ../sci-colorado_rpi
for image in **/*T2w*.nii.gz;do sct_image -i $image -setorient RPI;done
cd ../sci-paris_rpi/
for image in **/*T2w*.nii.gz;do sct_image -i $image -setorient RPI;done
```

**Comment by valosekj** [comment]
Experiments finished --> closing.

**Comment by naga-karthik** [comment]
Just to give more info on why this was closed -- we observed that there were only incremental gains (i.e. ~0.05 in Dice) with the pre-trained SC seg and fine-tuned lesion seg models compared to a single model that segmented both the SC and the lesion. Hence, in favour of having light model packaging, we proceeded with the single model.
